vyper/semantics/analysis/local.py

```diff
@@ -232,7 +232,10 @@ def visit_AnnAssign(self, node):
                 "Memory variables must be declared with an initial value", node
             )
 
-        type_ = type_from_annotation(node.annotation)
+        type_ = type_from_annotation(node.annotation, DataLocation.MEMORY)
+        if isinstance(type_, TupleT):
```

Comment:
this check is redundant now

Reply:
hm this is tricky. `DataLocation.MEMORY` is currently not in the `_invalid_locations` of `TupleT`, because the return type of a function can be a tuple which is also of `DataLocation.MEMORY` (see L315 of the amended `vyper.semantics.types.function.py` of this PR).

Reply:
https://github.com/vyperlang/vyper/blob/ba25230346b836291d67920934a07ae3578d99d7/vyper/semantics/types/function.py#L307-L319

Reply:
If we add `DataLocation.MEMORY` to `_invalid_locations` of `TupleT`, this contract will now fail because of the first exception instead of the second exception.

```
@internal
def _foo() -> (uint256, uint256):
    return 1, 2

@external
def foo():
    a: (uint256, uint256) = self._foo()
```
```
vyper.exceptions.InvalidType: (uint256, uint256) is not instantiable in DataLocation.MEMORY
  contract "contracts/tuple.vy:4", function "_foo", line 4:14 
       3 @internal
  ---> 4 def _foo() -> (uint256, uint256):
  ---------------------^
       5     return 1, 2
```
```
vyper.exceptions.VariableDeclarationException: Memory variables cannot be of tuple type
  contract "contracts/tuple.vy:9", function "foo", line 9:4 
       8 def foo():
  ---> 9     a: (uint256, uint256) = self._foo()
  -----------^
      10
```

Reply:
I think my original reply is incorrect since the return value can be in memory or storage. The location also may not be a reliable indicator of whether a type can be a return value of a function - tuples are not allowed in memory and storage, but can be a return value; hashmaps are allowed in storage but cannot be a return value. 

I have added a `_as_ret_val` attribute instead for this purpose.

Reply:
as discussed offline - we are just allowing tuples in memory (which is the same behavior as 0.3.7)